### ISSUE TITLE
Some Doubts 

### ISSUE THREAD

**Issue by amitbcp** [herobd/FUDGE#2]
Hi @herobd 

Thanks for the open sourcing the code. The Paper was a wonderful read, showcasing great work.
I had a couple of doubts, which can help me try this approach in a better informed way.

1. Can you specify the hardware(GPU & memory) you used for training so that I can estimate a similar time for me ?
2. As I understand, can we safely say that, we are doing node classification here as each node is getting classified into a semantic class ?
3. If yes, should we try to compare results with SROIE key value pair classification as well ?
4. For word-level FUDGE, if I want to bring in a custom dataset, what would be your suggestions and input ?
5. If I want to do a graph classification as well, would you recommend just concatenation node features and pass it through FC layers or somehow include edge information too ?

Thanks in advance !

**Comment by herobd** [comment]
1. I used a single GeForce GTX 1080 in training (8G). Using DistributedDataParallel or accumulating the gradient should be able to speed things up a bit. As mentioned in the README, the implementation is hard coded at a batch size of 1. However, you can increase the cropped window if you have more memory available on your card (also would need to adjust some memory helping parameterts like `percent_rel_to_keep` and `max_rel_to_keep`, and adjust `roi_batch_size` to be more efficient).

2. Yes we are doing node classification.

3. You definitely could. However, I suspect FUDGE won't perform well as it cannot read the text. SROIE is a bit of a different task as your trying to find specific text instances, rather than labeling the document as a whole.

4. If you are okay with the model being relient on external bounding box prediction, you should be able to adapt things pretty easily. I think you can throw our the idea of text lines entirely as we used text lines becuase their easier for the detector to find. You could train the detector at the word level; it won't do well, but that's okay becuase you'd just be replacing the predictions anyway (although Word-FUDGE does use the predicted boxes classes by assigning them to the GT word boxes). You'd then need to prevent the model from predicting a merge (which corrects oversegmentation) as you probably don't have text-line ground truth to train this. It should learn to group just fine from the words. If you do have text-line ground truth, I'm assuming you'd just use the same procedure as I used to train Word-FUDGE.

5. I would definitely include the edge features, even over the node ones. The edge features are initailized with the richest information and are responsible for the more difficult prediction task. You won't be able to just concatenate the features as there will be a variable number of edges and nodes for each graph. The standard way of doing this would be to just average all the node features, average all the edge features, concat these two vectors and pass this through a couple FC layers.

**Comment by amitbcp** [comment]
Thanks @herobd for your reply and insights.

Just one follow-up question. I get that in the paper we are trying to showcase how powerful visual features are and we are able to get comparable results with models like LayoutLM & BROS.

Since pertained LM currently are gigantic. But do you believe adding Text Embeddings to the nodes while initialisation would make any significant improvement in results. As I feel currently the FUDGE framework is trying to compete with one hand tied as it's not getting text features.

I was thinking, instead of using pretrained gigantic LM models, we can used simpler Character embeddings from Bi-LSTM or include a Bi-LSTM to get jointly trained for it as well.

What would be your take on that ?

**Comment by herobd** [comment]
I actually did try including word embeddings into the network as additional features, but was unable to get better results. Specifically, for the text of a given node I tried both
 - the average Word2Vec embedding (+FC layers)
 - running a byte-pair embedding through a Bi-LSTM that was jointly trained
and appending the result to the visual features for the node.

This was surprising and I only have speculative explinations for why. I have started working with transformers, and without pre-training they overfit to the FUNSD dataset very quickly. From an NLP perspective, FUNSD is quite small. However, it's larger if viewed from a visual perspective both becuase of the size of the images and the fact that image datasets can be smaller. As a results FUDGE is able learn stronger visual features.

Works like Carbonell et al. have had success with simpler word embeddings, but I'm guessing the information provided overlaps too much with what the CNNs can pick up on to be helpful.

FUDGE with the text features would train a bit faster, but would cap at the same performance. I tried to look at the results to see if there was a difference between the errors made by a model with or without the text embedding, but couldn't see anything obvious.

If you're interested in pursuing this, I could share my full code repo which still has these implemented (it's much messier).

**Comment by herobd** [comment]
The original point of the paper wasn't actually

**Comment by herobd** [comment]
Sorry didn't mean to close. Go ahead and ask more questions if you want

**Comment by amitbcp** [comment]
@herobd  thanks for sharing the text embedding results. I would definitely want to explore it, but I would first try to get comfortable with the current code and features to run it on my datasets and then bring in textual embeddings.

Thank you so much for your time ! :)
